uint8[] ids
int32[] positions
